Custom checkbox

lkleuver opened the pull request:
Change checkbox so it has custom styling to allow for error state.

part of #263

jorisleker commented on frontend/lib/ui/Checkbox/Checkbox.tsx:
```diff
@@ -12,13 +15,26 @@ interface CheckboxProps {
 //TODO: you can't style a border for a checkbox, currently outline is used as a workaround but the border radius doesnt match
```
I think this comment can be removed?

jorisleker reviewed (changes requested):
- The border-color of the unchecked unfocused (not in error state) checkbox is too dark ([Figma](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=2871-9247&node-type=FRAME&m=dev))
- Please use the focus (border+glow) we also use for buttons for the focus state of checkboxes (See [components](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=2906-35842&node-type=FRAME&m=dev) in Figma)
- When the warning is shown, the checkbox should have a error/600 red border. Currently it doesn't. ([Figma](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=3626-39567&node-type=FRAME&m=dev))
- The checkmark in design is a bit heavier then the current implementation ([Figma](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=3396-74006&node-type=FRAME&m=dev))
- The label element should have a top and border padding or margin of 1.5rem, so it is properly spaced to the alert and the submit button. 
- On small viewports, the checkbox's width is reduced, and the checkbox isn't square anymore. 

 ![Scherm­afbeelding 2024-09-03 om 13 04 22](https://github.com/user-attachments/assets/240a20da-40c3-4f0b-b6b9-debd64cef244)
![Scherm­afbeelding 2024-09-03 om 13 13 20](https://github.com/user-attachments/assets/b6273ad8-02b1-42f2-bc9e-8812410d30e6)

praseodym commented on frontend/e2e-tests/dom-to-db-tests/data-entry.e2e.ts:
```diff
@@ -448,7 +450,9 @@ test.describe("navigation", () => {
-      await votersVotesPage.acceptWarnings.click();
+
+      // eslint-disable-next-line playwright/no-force-option -- force option needed to click on hidden element
+      await votersVotesPage.acceptWarnings.click({ force: true, noWaitAfter: true });
```
Maybe we should also using be `.check` here? (cc @jschuurk-kr)

jschuurk-kr replied:
That would be better. I'd also prefer to have a `checkAcceptWarnings()` method in the page object instead of having to do this everywhere we want to accept warnings.

More importantly, locally the test fails every time at this point for me. Adding a `await expect(votersVotesPage.warning).toBeVisible();` before accepting the warning solved that.

lkleuver replied:
I've added a "checkAcceptWarnings" function to VotersVotesPgObj.ts
interestingly, there doesnt need to be a lint disable warning line here.

jschuurk-kr replied:
That's because the page object files are not included in the eslint `overrides` for Playwright.

Addressed in 772d6e770ea0bc382e08b31bb830e0e26d012c32.

praseodym reviewed (changes requested):
Could you remove the commits to `main` that are duplicated in this branch?

lkleuver commented:
> * The border-color of the unchecked unfocused (not in error state) checkbox is too dark ([Figma](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=2871-9247&node-type=FRAME&m=dev))
> * Please use the focus (border+glow) we also use for buttons for the focus state of checkboxes (See [components](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=2906-35842&node-type=FRAME&m=dev) in Figma)
> * When the warning is shown, the checkbox should have a error/600 red border. Currently it doesn't. ([Figma](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=3626-39567&node-type=FRAME&m=dev))
> * The checkmark in design is a bit heavier then the current implementation ([Figma](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=3396-74006&node-type=FRAME&m=dev))
> * The label element should have a top and border padding or margin of 1.5rem, so it is properly spaced to the alert and the submit button.
> * On small viewports, the checkbox's width is reduced, and the checkbox isn't square anymore.
> 

I've think I've addressed all issues, except the not showing the red border, I can't reproduce this.
If you can, could you check if the custom checkbox div has a class "has-error" ?

jschuurk-kr commented:
@lkleuver I don't get the red checkbox on the Voters and Votes page, only on the Differences page.

Lionqueen94 commented on frontend/lib/ui/Checkbox/Checkbox.module.css:
```diff
@@ -2,17 +2,55 @@
+
+  > div {
+    display: inline-block;
     margin-right: 8px;
+    flex: 0 0 20px;
+    width: 20px;
+    height: 20px;
+    border-radius: 6px;
```
Please convert the larger `px` amounts to `rem` (I made a lot of these changes in https://github.com/kiesraad/abacus/pull/303 for the codebase). See also `14px` on line 42.

lkleuver replied:
done

jschuurk-kr reviewed (changes requested):
If I check the checkbox, then make a change to the page that also results in a warning (same or different warning) and click "Volgende", the checkbox to accept warnings is already checked.

---

(edit) I wanted to write a test for this, but ran into a problem. I want to select the checkbox with:
```
screen.getByLabelText("Ik heb de aantallen gecontroleerd met het papier en correct overgenomen.")
```
but when I check for the checkbox for not being visible after editing the input, RTL claims the element is visible. I wonder if that's because the `hidden` is on the `<section>`, but that would mean RTL doesn't see children of a hidden parent as hidden, which seems weird.

This is not an issue with Playwright.

jschuurk-kr commented:
> (edit) I wanted to write a test for this, but ran into a problem. I want to select the checkbox with:
> 
> ```
> screen.getByLabelText("Ik heb de aantallen gecontroleerd met het papier en correct overgenomen.")
> ```
> 
> but when I check for the checkbox for not being visible after editing the input, RTL claims the element is visible. I wonder if that's because the `hidden` is on the `<section>`, but that would mean RTL doesn't see children of a hidden parent as hidden, which seems weird.

:point_up: This does seem to be the case,though. This fails:

```
      await user.clear(screen.getByTestId("blank_votes_count"));
      await user.type(screen.getByTestId("blank_votes_count"), "100");
      await user.tab();
      expect(screen.getByTestId("blank_votes_count"), "100").toHaveValue("100")
      
      await waitFor(() => {
        expect(screen.getByRole("checkbox", {
          name: "Ik heb de aantallen gecontroleerd met het papier en correct overgenomen.",
        })).not.tobeVisible();
      })
```
because according to RTL the element is visible, while the DOM output of the test does contain `<section class="row hidden">`.

lkleuver commented:
> * it is present in the document
> * it does not have its css property `display` set to `none`
> * it does not have its css property `visibility` set to either `hidden` or `collapse`
> * it does not have its css property `opacity` set to `0`
> * its parent element is also visible (and so on up to the top of the DOM tree)
> * it does not have the [`hidden`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/hidden) attribute
> * if `<details />` it has the `open` attribute

To summarize, styles aren't loaded by our testing library currently, we can either load them or focus RTL tests on checking for the correct classes/attributes.

Currently I've added the hidden attribute to the section that hides the checkbox. This works and circumvents loading styles. We could also check for the presence of "hidden" on the section, but this couples it a lot to the html structure.

jorisleker commented:
There is still the 'checkbox remains checked' bug. 

Reproduction steps:
- On voters and votes enter some numbers that will result in a warning (ie A100, D100, E90, F10, H100)
- press volgende
- check the 'accept warnings' checkbox, don't press volgende
- change your input to something that will trigger a warning (ie change E94, F6)
- the checkbox disappears
- press volgende. 
- you get a warning, the checkbox reappears and is checked (should be unchecked)

jorisleker reviewed (changes requested):
see last comment

lkleuver commented:
onChange didn't fire, changed it.
Might be a controlled input in the future.

jorisleker reviewed (approved):
retested the bug I found earlier, has been resolved!

lkleuver commented on frontend/app/component/form/voters_and_votes/VotersAndVotesForm.test.tsx:
```diff
@@ -368,6 +368,17 @@ describe("Test VotersAndVotesForm", () => {
+        const checkbox = screen.getByTestId("voters_and_votes_form_ignore_warnings");
+        expect(checkbox).toBeInTheDocument();
+        expect(checkbox).not.toBeVisible();
+      });
```
I've updated the test
